resource-thief.c

```diff
@@ -0,0 +1,361 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <signal.h>
+#include <fcntl.h>
+
+#include <gbm.h>
+#include <pool-buffer.h>
+#include <wayland-client-protocol.h>
+#include <linux-dmabuf-unstable-v1-client-protocol.h>
+
+#define POOL_SIZE 1024
+#define POINTS_PER_REGION 300
+#define DMABUF_WIDTH 64
+#define DMABUF_HEIGHT 64
+#define DMABUF_FORMAT GBM_FORMAT_XRGB8888
+
+static void handle_sigint(int sig) {
+	(void)sig;
+}
+
+/* Different types of resources to trick the compositor into exhausting */
+enum mode {
+	CONSUME_NOOP,
+	CONSUME_MAPS, /* wl_shm_pool */
+	CONSUME_MEMORY, /* wl_region */
+	CONSUME_FDS /* linux_dmabuf? */
```

Comment:
Maybe this should be named `CONSUME_DMABUF` to make it more clear DMA-BUFs are consumed? If the compositor raises its FD soft limit, the number of GL textures or something else might reach a limit sooner.

(Also, `CONSUME_MAPS` also consumes FDs.)

Reply:
> CONSUME_MAPS also consumes FDs.

`wl_shm_pool` only stores the mapping, not the fd: https://gitlab.freedesktop.org/wayland/wayland/-/blob/abcf1048e23525865c2ff43ce90bbdaa80524246/src/wayland-shm.c#L303 . Note also that `wl_shm_pool.resize` is handled with `mremap`, instead of making a new `mmap` call.

Reply:
Oh, interesting, I missed that. (Note, this might change and become system-dependent with BSD support.)

Reply:
>  (Note, this might change and become system-dependent with BSD support.)

DMABUF handling might also change in the future, and future protocols might provide better ways to use up memory or fds, so I've changed the command line argument to more closely reflect what object type is being made; i.e., dmabuf/region/shmpool.